include/deal.II/matrix_free/evaluation_template_factory.templates.h

```diff
@@ -0,0 +1,367 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2020 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+
+#ifndef dealii_matrix_free_evaluation_template_factory_templates_h
+#define dealii_matrix_free_evaluation_template_factory_templates_h
+
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/matrix_free/evaluation_kernels.h>
+#include <deal.II/matrix_free/evaluation_selector.h>
+#include <deal.II/matrix_free/evaluation_template_factory.h>
+#include <deal.II/matrix_free/fe_evaluation.h>
+
+#ifndef FE_EVAL_FACTORY_DEGREE_MAX
+#define FE_EVAL_FACTORY_DEGREE_MAX 6
+#endif
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace internal
+{
+  template <int degree, typename EvaluatorType, typename... Args>
+  bool
+  instantiation_helper_run(const unsigned int given_degree,
+                           const unsigned int n_q_points_1d,
+                           Args &... args)
+  {
+    if (given_degree == degree)
+      {
+        if (n_q_points_1d == degree + 1)
+          return EvaluatorType::template run<degree, degree + 1>(args...);
+        else if (n_q_points_1d == degree + 2)
+          return EvaluatorType::template run<degree, degree + 2>(args...);
+        else if (n_q_points_1d == degree)
+          return EvaluatorType::template run<degree, degree>(args...);
+        else
+          // slow path
+          return EvaluatorType::template run<-1, 0>(args...);
+      }
+    else if (degree < FE_EVAL_FACTORY_DEGREE_MAX)
+      return instantiation_helper_run<
+        (degree < FE_EVAL_FACTORY_DEGREE_MAX ? degree + 1 : degree),
+        EvaluatorType>(given_degree, n_q_points_1d, args...);
+    else
+      // slow path
+      return EvaluatorType::template run<-1, 0>(args...);
+  }
```

Comment:
This is the interesting function that recursively goes into all other degrees. If we wanted more quadrature formulas we would add them here.

Reply:
I think here is one quadrature formula missing which we need? Over-integration?

Reply:
I was a bit split about over-integration, but yes, I added it now. It only adds code for degrees 4 to 6 because `3*degree/2+1` gets `degree+1` for degree 1 and `degree+2` for degrees 2 and 3.

Reply:
Thanks!